static/js/composebox_typeahead.js

```diff
@@ -189,78 +185,39 @@ function handle_keydown(e) {
         const on_pm = target_sel === "#private_message_recipient";
         const on_compose = target_sel === "#compose-textarea";
 
-        if (on_stream || on_topic || on_pm) {
-            // For Enter, prevent the form from submitting
-            // For Tab, prevent the focus from changing again
-            e.preventDefault();
-        }
-
-        // In the compose_textarea box, preventDefault() for Tab but not for Enter
-        if (on_compose && code !== 13) {
-            e.preventDefault();
-        }
-
-        if (on_stream) {
-            nextFocus = "#stream_message_recipient_topic";
-        } else if (on_topic) {
-            if (code === 13) {
+        if (on_compose && code === 13) {
+            if (exports.should_enter_send(e)) {
                 e.preventDefault();
+                if (!$("#compose-send-button").prop("disabled")) {
+                    $("#compose-send-button").prop("disabled", true);
+                    compose.finish();
+                }
+                return;
             }
-            nextFocus = "#compose-textarea";
-        } else if (on_pm) {
-            nextFocus = "#compose-textarea";
-        } else if (on_compose) {
-            if (code === 13) {
-                nextFocus = false;
-            } else {
-                nextFocus = "#compose-send-button";
-            }
-        } else {
-            nextFocus = false;
-        }
-
-        // If no typeaheads are shown...
-        if (
-            !(
-                $("#stream_message_recipient_topic").data().typeahead.shown ||
-                $("#stream_message_recipient_stream").data().typeahead.shown ||
-                $("#private_message_recipient").data().typeahead.shown ||
-                $("#compose-textarea").data().typeahead.shown
-            )
-        ) {
-            // If no typeaheads are shown and the user is tabbing from the message content box,
-            // then there's no need to wait and we can change the focus right away.
-            // Without this code to change the focus right away, if the user presses Enter
-            // before they fully release the Tab key, the Tab will be lost.  Note that we don't
-            // want to change focus right away in the private_message_recipient box since it
-            // takes the typeaheads a little time to open after the user finishes typing, which
-            // can lead to the focus moving without the autocomplete having a chance to happen.
```

Comment:
Do you understand whether we're regression behavior here for folks hitting `Tab+Enter` very quickly once this PR is merged?  The old comment here suggests we might be, but it's quite old.

Also, is there an explanation for why we no longer need to check if typeaheads are shown in this logic?

Reply:
I did a round of testing and wasn't able to cause `Tab+Enter` to fail to send in any of my tests, including where I conspicuously held down `Tab` until after I hit `Enter`.

Reply:
>Do you understand whether we're regression behavior here for folks hitting Tab+Enter very quickly once this PR is merged? The old comment here suggests we might be, but it's quite old.

I think this was only a problem previously because `nextFocus` was set to `false` in `keydown` under some circumstances (which I no longer do).

>Also, is there an explanation for why we no longer need to check if typeaheads are shown in this logic?

It should never happen that a recipient typeahead is open while the compose textarea has focus and the composebox typeaheads stop the event propagation, so that this code path isn't reached anyway.